### ISSUE TITLE
Provide output from git command failures

### ISSUE THREAD

**Issue by supercow** [puppetlabs/r10k#96]
When a git command run by r10k fails for any reason, the full command line and an exit code 128 is printed to the console, but the details of the error are lost, forcing the user to manually re-run the command to see what the issue was. It would be great if error messages from git were captured and printed to STDOUT when there is a failure.

**Comment by adrienthebo** [comment]
This should be partially fixed in 1.2.0 with the fix in https://github.com/adrienthebo/r10k/blob/master/lib/r10k/git/cache.rb#L65-L70, although I could use more information as to what should be logged. Would you prefer to see all of stderr logged on error?

**Comment by supercow** [comment]
+1 for dumping STDERR. I just want to know exactly what failed so I don't need to copy paste git commands and manually run them.

**Comment by adrienthebo** [comment]
Tenatively targeting this at 1.4.0.

**Comment by adrienthebo** [comment]
I've reworked error logging to include more context. This is what it currently looks like:

Attempting to fetch an inaccessible repo for caching:

```
[R10K::Util::Attempt - ERROR] Couldn't update git cache for git@github.com:rtyler/jenkins-keys.git
Original exception:
Command exited with non-zero exit code:
Command: git clone --mirror git@github.com:rtyler/jenkins-keys.git /home/adrien/.r10k/git/git@github.com-rtyler-jenkins-keys.git
Stderr:
Cloning into bare repository '/home/adrien/.r10k/git/git@github.com-rtyler-jenkins-keys.git'...
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
Exit code: 128
```

Installing a module with bad metadata:

```
[R10K::Util::Attempt - ERROR] Command exited with non-zero exit code:
Command: puppet module --modulepath /home/adrien/development/r10k/testing/jenkins-infra/environments/staging/modules install --version=1.0.2 --force rtyler/groovy
Stdout:
Notice: Preparing to install into /home/adrien/development/r10k/testing/jenkins-infra/environments/staging/modules ...
Notice: Downloading from https://forgeapi.puppetlabs.com ...
Stderr:
Warning: Cannot consider release rtyler-groovy-1.0.2: Malformed dependency: nanliu/staging. Exception was: Invalid 'version_range' field in metadata.json: Unparsable version range: "~> 0.4.0"
Warning: Cannot consider release rtyler-groovy-1.0.0: Malformed dependency: puppetlabs/stdlib. Exception was: Invalid 'version_range' field in metadata.json: Unparsable version range: "~> 3.2.1"
Error: Could not install 'rtyler-groovy' (v1.0.2)
  No releases matching '1.0.2' are available from https://forgeapi.puppetlabs.com
Exit code: 1
```

Is this too wordy or is it better to print this much information on error cases?

**Comment by binford2k** [comment]
+1

**Comment by joshbeard** [comment]
:+1:

**Comment by adrienthebo** [comment]
Logging revised in 33cdd81.

**Comment by adrienthebo** [comment]
@supercow @joshbeard @binford2k thanks for weighing in on this!
